packages/lesswrong/server/resolvers/bookmarksFeedResolver.ts

```diff
@@ -0,0 +1,165 @@
+  BookmarksFeed: async (_root: void, args: any, context: ResolverContext): Promise<BookmarksFeedQueryResult> => {
+    const {limit = 20, cutoff, offset = 0} = args;
+    const {currentUser} = context;
+    
+    if (!currentUser) {
+      throw new Error("Must be logged in to view bookmarks feed");
+    }
+
+    const selector: MongoSelector<DbBookmark> = {
+      userId: currentUser._id,
+      cancelled: false,
+    };
+    if (cutoff) {
+      selector.lastUpdated = {$lt: cutoff}; 
+    }
+
+    const bookmarks = await Bookmarks.find(selector, {
+      sort: {lastUpdated: -1},
+      limit: limit + 1, 
+      ...(offset && !cutoff ? { skip: offset } : {}) 
+    }).fetch();
+
+    const hasMore = bookmarks.length > limit;
+    const resultsBookmarks = hasMore ? bookmarks.slice(0, limit) : bookmarks;
+    const nextCutoff = hasMore ? (resultsBookmarks[resultsBookmarks.length - 1]?.lastUpdated ?? new Date()) : null;
+
+    const postBookmarkIds: string[] = [];
+    const commentBookmarkIds: string[] = [];
+    const bookmarkMap = new Map<string, DbBookmark>();
+
+    resultsBookmarks.forEach(bookmark => {
+      if (bookmark.documentId) {
+        bookmarkMap.set(bookmark.documentId, bookmark);
+        if (bookmark.collectionName === 'Posts') {
+          postBookmarkIds.push(bookmark.documentId);
+        } else if (bookmark.collectionName === 'Comments') {
+          commentBookmarkIds.push(bookmark.documentId);
+        }
+      }
+    });
+
+    const [posts, comments] = await Promise.all([
+      postBookmarkIds.length ? loadByIds(context, "Posts", postBookmarkIds) : Promise.resolve([]),
+      commentBookmarkIds.length ? loadByIds(context, "Comments", commentBookmarkIds) : Promise.resolve([])
+    ]);
+    
+    const commentPostIds = filterNonnull(comments).map(c => c.postId);
+    const validCommentPostIds = filterNonnull(commentPostIds);
+    const commentPosts = validCommentPostIds.length ? await loadByIds(context, "Posts", uniq(validCommentPostIds)) : [];
+    const commentPostsById = new Map(filterNonnull(commentPosts).map(p => [p._id, p]));
+
+    const feedItems: InternalFeedItem[] = [];
+
+    filterNonnull(posts).forEach(post => {
+      const bookmark = bookmarkMap.get(post._id);
+      if (!bookmark) return;
+
+      const postMetaInfo: FeedPostMetaInfo = {
+        sources: ["bookmarks"],
+        displayStatus: "expanded" 
+      };
+      const feedPostData: FeedPostResolverType = {
+         _id: post._id,
+         post: post,
+         postMetaInfo: postMetaInfo
+      };
+      feedItems.push({
+        type: "feedPost",
+        feedPost: feedPostData,
+        feedCommentThread: null,
+        sortKey: bookmark.lastUpdated ?? new Date()
+      });
+    });
+
+    filterNonnull(comments).forEach(comment => {
+      const bookmark = bookmarkMap.get(comment._id);
+      if (!bookmark || !comment.postId) return;
+
+      const commentMetaInfos: { [commentId: string]: FeedCommentMetaInfo } = {
```

Comment:
The idiomatic type annotation here is `Record<string, FeedCommentMetaInfo>`.   (I'm sorta confused by why we have a single field key with the comment id, but my guess is that this is conforming a pre-existing interface?)

Reply:
yeah, the way it's set up comments display as part of threads, and threads typically have multiple comments, but for bookmarks right now just wrapping individualy comments